crates/proto/src/error.rs

```diff
@@ -260,28 +260,32 @@ impl fmt::Display for ProtoError {
 }
 
 impl From<ProtoErrorKind> for ProtoError {
+    #[cold]
```

Comment:
Do these `cold` annotations have an observable effect on your benchmarks, by themselves?

Reply:
I do not have anything wildly sophisticated for benchmarks. I'm running this under a production workload for a few minutes and looking what `perf` tells me about time spent in `Name::read`.

Per the microbenchmarks I added in this PR, the cold annotation is responsible for no more than a decrease of 150 out of 4200 ns/iter. That is small compared to the other win of reducing allocation in `Name`, and could probably be obviated by cleaning up the error reporting in this module (lots of the parsing functions return a string literal as an error, that could be packed into 1 byte). If we'd rather revisit errors more holistically in another PR, I'd be happy to send one.

Reply:
> could probably be obviated by cleaning up the error reporting in this module (lots of the parsing functions return a string literal as an error, that could be packed into 1 byte). If we'd rather revisit errors more holistically in another PR, I'd be happy to send one.

That sounds good to me.